### PR TITLE
UPSTREAM: 14881: fix delta fifo & various fakes for go1.5.1

### PR DESCRIPTION
**Pull request by soltysh**
This is re-submission of the upstream patch which must be included before next rebase lands. This was previously submitted by @deads2k in #5339, this adds @bparees patch on top of it. 

Ben, frankly said I'm moderately happy with that solution of yours, but it works. I'd rather us switch to using `framework.Informer` for that, but that keeps internal cache you don't want in return (similarly to  [this](https://github.com/openshift/origin/pull/6230/files#diff-54ebade8eae1319de4596fd6f5f1f612R354)).

### CODE REVIEW

File: pkg/build/controller/factory/factory.go [hunk 1]
```diff
@@ -206,11 +206,29 @@ func (factory *BuildPodControllerFactory) Create() controller.RunnableController
...
+// returns no items for list.  This forces the DeltaFIFO queue
+// to always queue delete events it receives from etcd.  Our
+// client will properly handle duplicated events and this is more
+// efficient than maintaining a local cache of all the build pods
+// so the DeltaFIFO can perform a proper diff.
+type keyListerGetter struct {
+	client osclient.Interface
+}
+
+func (keyListerGetter) ListKeys() []string {
```
Comment (mfojtik) [inline pkg/build/controller/factory/factory.go]: godoc ;-)

**Comment by soltysh** [comment]
[test][extended:core(builds)]

**Comment by bparees** [comment]
moderately happy?  you don't sound moderately happy about.  but luckily i don't care :)

**Comment by soltysh** [comment]
This doesn't seen related:
```
[ ERR] Gave up waiting for: 'oc get rc/failing-dc-1 --template={{.metadata.annotations}} | grep openshift.io/deployment.phase:Failed'
```
re-[test][extended:core(builds)]

**Comment by soltysh** [comment]
Now this is mean, got another flake from #6232.
re-[test][extended:core(builds)]

**Comment by soltysh** [comment]
Tests are working ok, the extended test failures are known problems (#6215,#5995). @bparees @deads2k any objections for merging this?

**Comment by bparees** [comment]
@mfojtik still wants you to godoc some stuff.  otherwise i have no objection.

**Comment by soltysh** [comment]
To make golint and @mfojtik happy, in that order ;)

**Comment by mfojtik** [comment]
image

**Comment by soltysh** [comment]
@mfojtik you're happy, but the test are not ;( still I got green for merge and these are flakes so let's do it.
[merge]

**Comment by soltysh** [comment]
I thought it's already in, needed for the rebase.
[merge]

**Pull request merged**